gwctl/pkg/resources/backends/backends.go

```diff
@@ -0,0 +1,282 @@
+			result[gatewayRef], err = policymanager.MergePoliciesOfSameHierarchy(result[gatewayRef], policies)
+			if err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	// Step 5: Loop through all Gateways and merge the Backend and
+	// Backend-namespace specific policies. Note that this needs to be done
+	// separately from Step 4 i.e. we can't have this loop within Step 4 itself.
+	// This is because we first want to merge all policies of the same-hierarchy
+	// together and then move to the next hierarchy of Backend and
+	// Backend-namespace.
+	for gatewayRef := range result {
+		// Merge all hierarchial policies.
+		result[gatewayRef], err = policymanager.MergePoliciesOfDifferentHierarchy(result[gatewayRef], backendNamespacePoliciesByKind)
+		if err != nil {
+			return nil, err
+		}
+
+		result[gatewayRef], err = policymanager.MergePoliciesOfDifferentHierarchy(result[gatewayRef], backendPoliciesByKind)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return result, nil
+}
+
+func httpRoutesForBackend(ctx context.Context, params *types.Params, backend unstructured.Unstructured) ([]gatewayv1beta1.HTTPRoute, error) {
+	allHTTPRoutes, err := httproutes.List(ctx, params, "")
+	if err != nil {
+		return nil, err
+	}
+
+	var filteredHTTPRoutes []gatewayv1beta1.HTTPRoute
+	for _, httpRoute := range allHTTPRoutes {
+		found := false
+
+		for _, rule := range httpRoute.Spec.Rules {
+			for _, backendRef := range rule.BackendRefs {
+				if *backendRef.Group != gatewayv1beta1.Group(backend.GroupVersionKind().Group) {
+					continue
+				}
+				if *backendRef.Kind != gatewayv1beta1.Kind(backend.GroupVersionKind().Kind) {
+					continue
+				}
+				if backendRef.Name != gatewayv1beta1.ObjectName(backend.GetName()) {
+					continue
+				}
+				var ns string
+				if backendRef.Namespace != nil {
+					ns = string(*backendRef.Namespace)
+				}
+				if ns == "" {
+					ns = httpRoute.GetNamespace()
+				}
+				if ns != backend.GetNamespace() {
+					continue
+				}
+				found = true
+				break
+			}
+			if found {
+				break
+			}
+		}
+
+		if found {
+			filteredHTTPRoutes = append(filteredHTTPRoutes, httpRoute)
+		}
+	}
+
+	return filteredHTTPRoutes, nil
+}
+
+type describeView struct {
+	Group                    string                                                        `json:",omitempty"`
+	Kind                     string                                                        `json:",omitempty"`
+	Name                     string                                                        `json:",omitempty"`
+	Namespace                string                                                        `json:",omitempty"`
+	DirectlyAttachedPolicies []policymanager.ObjRef                                        `json:",omitempty"`
+	EffectivePolicies        map[string]map[policymanager.PolicyCrdID]policymanager.Policy `json:",omitempty"`
+}
+
+func PrintDescribeView(ctx context.Context, params *types.Params, backendsList []unstructured.Unstructured) {
```

Comment:
Non-blocking nit: To simplify the use of this code as a library it may be best to separate code into different packages. For example, one package or set of packages intended to be used as a lib, and a separate package that is more unique to gwctl inputs and outputs (like this describe view here).

I'm not really sure what the ideal code structure is here yet, so this may be worth filing a follow up issue for more discussion once this one merges.

Reply:
I definitely resonate with your thoughts here. There's several improvements that could be made to make this more usable as a library. I created https://github.com/kubernetes-sigs/gateway-api/issues/2435 to track the work for this.

Reply:
Refactored things around. This should be slightly better than the previous iteration. Function names need some refinement but the packages `effectivepolicy` and `policymanager` provide constructs that can be used as a library and are independent from the CLI part.